### PR TITLE
rectMoreOrLess equals, prep for 64bit rects

### LINKED ISSUE #30920: "Change Rect internal representation from Float32List to Float64List (#8524)" causes Rect comparisons to fail
Sample of failing test ([full logs here](https://api.cirrus-ci.com/v1/task/5135058510807040/logs/test.log)):

```
04:04 +3164 ~31 -6: /tmp/flutter sdk/packages/flutter/test/material/scaffold_test.dart: Scaffold and extreme window padding                                                                            
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: Rect:<Rect.fromLTRB(36.0, 255.0, 113.0, 332.0)>
  Actual: Rect:<Rect.fromLTRB(36.0, 255.0, 113.0, 332.0)>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///tmp/flutter%20sdk/packages/flutter/test/material/scaffold_test.dart:765:5)
<asynchronous suspension>
#5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:108:25)
<asynchronous suspension>
#6      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:589:19)
<asynchronous suspension>
#9      TestWidgetsFlutterBinding._runTest (package:flutter_test/src/binding.dart:573:14)
#10     AutomatedTestWidgetsFlutterBinding.runTest.<anonymous closure> (package:flutter_test/src/binding.dart:920:24)
#16     AutomatedTestWidgetsFlutterBinding.runTest (package:flutter_test/src/binding.dart:917:15)
#17     testWidgets.<anonymous closure> (package:flutter_test/src/widget_tester.dart:106:22)
#18     Declarer.test.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:168:27)
<asynchronous suspension>
#19     Invoker.waitForOutstandingCallbacks.<anonymous closure> (package:test_api/src/backend/invoker.dart:250:15)
<asynchronous suspension>
#24     Invoker.waitForOutstandingCallbacks (package:test_api/src/backend/invoker.dart:247:5)
#25     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:166:33)
#30     Declarer.test.<anonymous closure> (package:test_api/src/backend/declarer.dart:165:13)
<asynchronous suspension>
#31     Invoker._onRun.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/invoker.dart:400:25)
<asynchronous suspension>
#45     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:382:19)
#46     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
#47     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)
(elided 28 frames from class _FakeAsync, package dart:async, package dart:async-patch, and package
stack_trace)

This was caught by the test expectation on the following line:
  file:///tmp/flutter%20sdk/packages/flutter/test/material/scaffold_test.dart line 765

The test description was:
Scaffold and extreme window padding
════════════════════════════════════════════════════════════════════════════════════════════════════
```

Bisected failing engine roll to the following commit: https://github.com/flutter/engine/commit/0b36d3e2f0235f25dc5b5657bd2e594fe7414aee

See flutter/engine#8545 for more details.

### PR DESCRIPTION
**Pull request by dnfield**
## Description

https://github.com/flutter/engine/pull/8524 had to be reverted because the increased precision fixed some issues, but caused our tests to fail for others.

In paritcular, FloatingActionButton animations involve a transform that gets very slightly translated. After going through various matrix multiplications, there's a slight loss of precision now that's captured by a double backed `Rect`.  That's resolved with a new matcher called `rectMoreOrLessEquals`, which works like `moreOrLessEquals`.

The semantics change is because the rect also has more precision, and when that extra precision is pumped through to `SizedBox` it causes a failure.  I'm hoping @yjbanov or @goderbauer can validate that this is safe - it seems so to me but I may be missing a bug in the semantics traversal algorithm.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/30920
Allows us to reland flutter/engine#8524 (or something similar to it).

## Tests

I added the following tests:

New tests for `rectMoreOrLessEquals`.
Updated tests in preparation for increased precision in `Rect`s.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_test/lib/src/matchers.dart [hunk 1]
```diff
@@ -222,6 +222,21 @@ Matcher moreOrLessEquals(double value, { double epsilon = 1e-10 }) {
...
+///
+/// Two values are considered equal if the difference between them is within
+/// 1e-10 of the larger one. This is an arbitrary value which can be adjusted
+/// using the `epsilon` argument. This matcher is intended to compare floating
+/// point numbers that are the result of different sequences of operations, such
+/// that they may have accumulated slightly different errors.
+///
+/// See also:
+///
+///  * [moreOrLessEquals], which is for [double]s.
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter_test/lib/src/matchers.dart]: Also, why can't we use `within`?

[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter_test/lib/src/matchers.dart]: It looks like we could, but it seems confusing.  reading something like `expect(find.byKey(fabKey).getRect(), within(Rect.fromLTRB(...))` makes me think we expect the fab rect to be smaller - whereas`rectMoreOrLessEquals` reads a little more clearly IMHO

[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter_test/lib/src/matchers.dart]: SGTM

[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter_test/lib/src/matchers.dart]: But let's make sure the two are more or less (:wink:) consistent.

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter_test/lib/src/matchers.dart]: Can `rectMoreOrLessEquals` internally just defer to within to ensure they are consistent?

[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter_test/lib/src/matchers.dart]: That works for me.  I'm trying to figure out why the semantics traversal test is behaving differently on different platforms right now.

**Comment by dnfield** [comment]
We will have to update the `semantics_traversal_test` with the roll.  I can't quite get it to work in both 64 and 32 bit precision, and it doesn't seem worth it to spend more effort on that when we can just fix it in the roll.

**Comment by Hixie** [comment]
LGTM, see also https://github.com/flutter/flutter/pull/31333

**Pull request merged**